[inductor] Defer memory operation lowering to wrapper

int3 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111402

Right now, memory ops are being lowered to strings partly in
scheduler.codegen() and partly in wrapper.codegen(). But that makes
static memory planning (which is done entirely in `wrapper.codegen()`)
difficult to implement as information is "lost" by that point.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

int3 committed:
[inductor] Defer memory operation lowering to wrapper

Right now, memory ops are being lowered to strings partly in
scheduler.codegen() and partly in wrapper.codegen(). But that makes
static memory planning (which is done entirely in `wrapper.codegen()`)
difficult to implement as information is "lost" by that point.

[ghstack-poisoned]

int3 commented:
`codegen_inplace_reuse()` is still being called from `scheduler.codegen()`, but I guess reuse decisions don't affect memory planning decisions as it is currently implementated. I'm wondering if it would make sense to move that to wrapper.codegen() too though

int3 committed:
Update on "[inductor] Defer memory operation lowering to wrapper"


Right now, memory ops are being lowered to strings partly in
scheduler.codegen() and partly in wrapper.codegen(). But that makes
static memory planning (which is done entirely in `wrapper.codegen()`)
difficult to implement as information is "lost" by that point.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -853,6 +855,10 @@ def make_allocation(self, name, device, dtype, shape, stride):
             )
 
     def make_buffer_free(self, buffer):
+        name = buffer.get_name()
+        if name in self.freed:
```
Don't we need to do this on the last free, not the first one?

jansel reviewed (changes requested):
Tests are failing

int3 commented:
This wasn't meant for review yet, that's why I marked it as a draft... should I be taking off the reviewers as well? This isn't the first time I've had a draft reviewed :/

int3 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -853,6 +855,10 @@ def make_allocation(self, name, device, dtype, shape, stride):
             )
 
     def make_buffer_free(self, buffer):
+        name = buffer.get_name()
+        if name in self.freed:
```
sorry, not sure I understand this comment. What is the first/last free you're referring to here?

int3 committed:
Update on "[inductor] Defer memory operation lowering to wrapper"


Right now, memory ops are being lowered to strings partly in
scheduler.codegen() and partly in wrapper.codegen(). But that makes
static memory planning (which is done entirely in `wrapper.codegen()`)
difficult to implement as information is "lost" by that point.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Defer memory operation lowering to wrapper"


Right now, memory ops are being lowered to strings partly in
scheduler.codegen() and partly in wrapper.codegen(). But that makes
static memory planning (which is done entirely in `wrapper.codegen()`)
difficult to implement as information is "lost" by that point.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

int3 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1531,6 +1531,10 @@ def can_cache_buffer_in_thread_local(self, buffer):
+        name = buffer.get_name()
+        if name in self.freed:
+            return ""
+        self.freed.add(name)
```
I've removed these entirely. I'm not sure why this change was introduced in your initial diff... but things seem to work regardless

int3 committed:
Update on "[inductor] Defer memory operation lowering to wrapper"


Right now, memory ops are being lowered to strings partly in
scheduler.codegen() and partly in wrapper.codegen(). But that makes
static memory planning (which is done entirely in `wrapper.codegen()`)
difficult to implement as information is "lost" by that point.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

int3 commented:
@pytorchbot merge
